[WIP] Contributors along with their avatars showing up on multi-repository view.

Rishabh570 opened the pull request:
Fixes #29

jywarren commented:
Awesome! Could you share a screenshot?

On Thu, Oct 4, 2018, 9:36 AM Rishabh Rawat <notifications@github.com> wrote:

> Fixes #29 <https://github.com/publiclab/community-toolbox/issues/29>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/publiclab/community-toolbox/pull/54
> Commit Summary
>
>    - Contributors along with their avatars showing up on multi-repository
>    view.
>
> File Changes
>
>    - *M* dist/community-toolbox.js
>    <https://github.com/publiclab/community-toolbox/pull/54/files#diff-0>
>    (9638)
>    - *M* index.html
>    <https://github.com/publiclab/community-toolbox/pull/54/files#diff-1>
>    (3)
>    - *M* src/community-toolbox.js
>    <https://github.com/publiclab/community-toolbox/pull/54/files#diff-2>
>    (87)
>
> Patch Links:
>
>    - https://github.com/publiclab/community-toolbox/pull/54.patch
>    - https://github.com/publiclab/community-toolbox/pull/54.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/community-toolbox/pull/54>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ3K-NuJ7VJCvhj7cy9EQW_BxYucwks5uhg7sgaJpZM4XIH_X>
> .
>

Rishabh570 commented:
@jywarren I open this *work in progress* PR. 
There are some small things to be fixed in this, but most of the work is done.
What I've done is, fetch all the repos and then store all the **unique** contributors from all these repos (currently first 30 repos), and store them in **localStorage**, so that with a page reload we don't hit Github API redundantly and use **localStorage** instead...

Once I fix some remaining things, I'll post a screenshot of it :)

jywarren commented:
Great, very eager to see it!

On Thu, Oct 4, 2018, 9:44 AM Rishabh Rawat <notifications@github.com> wrote:

> @jywarren <https://github.com/jywarren> I open this *work in progress* PR.
> There are some small things to be fixed in this, but most of the work is
> done.
> What I've done is, fetch all the repos and then store all the *unique*
> contributors from all these repos (currently first 30 repos), and store
> them in *localStorage*, so that with a page reload we don't hit Github
> API redundantly and use *localStorage* instead...
>
> Once I fix some remaining things, I'll post a screenshot of it :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/community-toolbox/pull/54#issuecomment-427023995>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJyYYFt3P6GM6UqHuksvm0w7-xlwxks5uhhCjgaJpZM4XIH_X>
> .
>

jywarren commented:
👍 This is one of my top bugs I'd like to see fixed, so thank you in advance!!!

Rishabh570 commented on src/community-toolbox.js:
```diff
@@ -39,6 +39,136 @@ CommunityToolbox = function CommunityToolbox(org, repo) {
+    var testRepos = repos.splice(0,2);
+    testRepos.map((repo, i) => {
+      var repoContributors = fetchRepoContributors(org, repo);
+      console.log("total contributors for repo ", i, "are : ", repoContributors ,repoContributors.length, '\n')
```
Here, `repoContributors.length` is 0 while `repoContributors` has objects in it...What is the reason for this behaviour??

Rishabh570 commented:
@jywarren I was stuck onto something (implementing promises, using async/await) but none seem to work out...For now I get it to work with the help of `setTimeout` function which also achieves what we want except that opening `https://code.publiclab.org` for the first will take a while to load contributors (approx 20sec) but after that it will never hit Github API for loading contributors again (as long as localstorage is not cleared out)...


![screenshot from 2018-10-05 17-08-40](https://user-images.githubusercontent.com/25483260/46533523-87104080-c8c2-11e8-8ec7-51ac6137edea.png)


Please take a look :)

jywarren commented:
> take a while to load contributors (approx 20sec)

I think that's OK -- and i'm happy to look at this now but keep in mind
that wherever possible we should try to solve things one function at a time
-- it makes merging PRs much easier and makes for better modularized code.
Thank you though!

On Fri, Oct 5, 2018 at 7:48 AM Rishabh Rawat <notifications@github.com>
wrote:

> @jywarren <https://github.com/jywarren> I was stuck onto something
> (implementing promises, using async/await) but none seem to work out...For
> now I get it to work with the help of setTimeout function which also
> achieves what we want except that opening https://code.publiclab.org for
> the first will take a while to load contributors (approx 20sec) but after
> that it will never hit Github API for loading contributors again (as long
> as localstorage is not cleared out)...
>
> [image: screenshot from 2018-10-05 17-08-40]
> <https://user-images.githubusercontent.com/25483260/46533523-87104080-c8c2-11e8-8ec7-51ac6137edea.png>
>
> Please take a look :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/community-toolbox/pull/54#issuecomment-427338363>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJxTvFun5gQcKl4T140vK_ClobgEbks5uh0bjgaJpZM4XIH_X>
> .
>

jywarren commented on src/community-toolbox.js:
```diff
@@ -39,6 +39,130 @@ CommunityToolbox = function CommunityToolbox(org, repo) {
+           })
+
+
+  return new Promise((resolve, reject) => {
```
so you did end up using promises? cool!

Rishabh570 replied:
Yeah, I used them but things didn't happen the way I thought, so I ended up putting up `setTimeout` everywhere...(async/await are not allowed by grunt-browserify, it is giving error while `grunt build`)

jywarren commented on src/community-toolbox.js:
```diff
@@ -39,6 +39,130 @@ CommunityToolbox = function CommunityToolbox(org, repo) {
+      res.map((item, i) => {
+        repos[i] = item.name;
+      });
+      // Stores all public lab repos to localstorage (only first 30 repos yet)
```
This could become a new follow-up issue 👍

Rishabh570 replied:
Yes, currently fetching publiclab's repos only gives the repos listed on 1st page...

jywarren commented on src/community-toolbox.js:
```diff
@@ -39,6 +39,130 @@ CommunityToolbox = function CommunityToolbox(org, repo) {
+            });
+            }
+            // Stores all unique contributors info to localstorage
+            localStorage.setItem('contributors', JSON.stringify(myArr));
```
Interesting! When would this then get flushed and updated? Could we do once daily? And could we implement this via a `flushLocalStorageAllContributors` or something... so we could manually refresh it with a 🔄  button or something?

Rishabh570 replied:
Currently, the only way to request the API again for contributors is by manually flushing localStorage...but this can easily done, so no worries :)

jywarren commented:
Also, I'd love to have contributors listed with `@` so they can be quickly copied and pasted.

Rishabh570 commented:
Updating it :smile_cat:

jywarren commented:
Maybe we should store the date in the localstorage too, so we know whether
to flush it (if it's older than 1 day, for example?)

On Fri, Oct 5, 2018 at 1:07 PM Rishabh Rawat <notifications@github.com>
wrote:

> Updating it 😸
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/community-toolbox/pull/54#issuecomment-427434316>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABfJ7Y5TCN5Y32ByzCRda9PBqCdscqWks5uh5F9gaJpZM4XIH_X>
> .
>

Rishabh570 commented:
@jywarren I've added the logic for removing contributors data from localStorage after every single day...please take a look :)

jywarren commented:
This looks great. Did you want to try to push this into your gh-pages branch so we can try it before merging, OR post a screenshot gif using the `peek` or `licecap` programs?

We should look ahead to tests as well so we don't have to be so tentative :-)

Thank you!!!! 👍

Rishabh570 commented:
Hi @jywarren , sorry for delay, actually I made gif 3 times but it gets bigger than 10mb everytime and I keep hitting the API limits regularly while testing...

I'll create another gif as my API quota gets refilled :sweat_smile:

Rishabh570 commented:
@jywarren Here is the working GIF,

![ezgif com-optimize](https://user-images.githubusercontent.com/25483260/46572249-222b1800-c9a0-11e8-98a5-c3a4ced0e042.gif)

jywarren commented:
Ok awesome. Why 20 seconds, precisely? Just waiting for all results to come in? Could we display them as they arrive? It does seem a little long to wait I guess with no feedback. 

Also, perhaps if we're waiting we should add a spinner icon - fontawesome has a `<i class="fa fa-spinner fa-spin"></i>` icon class you could use. What do you think?

Rishabh570 commented:
@jywarren Finally, removed `setTimeout` from everywhere...implemented promises instead :sweat_smile: 
I'll be sharing a working GIF (as my API quota gets refilled)

Rishabh570 commented:
@jywarren Here it is, it doesn't take 20seconds anymore :joy: Probably 2-3 seconds...

![working_promises](https://user-images.githubusercontent.com/25483260/46584094-d5693f00-ca7c-11e8-97a8-17ea12de1406.gif)

jywarren commented:
This is fantastic. Do you think it's ready to merge?

Rishabh570 commented:
Yes, only if we don't need spinner icon for now...

Rishabh570 commented:
@jywarren what do you think...do we need a spinner icon?

jywarren commented:
Let's do spinner in a follow up. Would you mind opening an issue? 

Merging! I'll publish soon :-)))

Great work!!!!

jywarren commented:
Hi, @Rishabh570  -- just another thought here -- does this properly remember to show individual contributors when you go to:

http://code.publiclab.org

after you first go to http://code.publiclab.org#r=all ?

Rishabh570 commented:
@jywarren Yes, because there are the separate functions for both of the views...I will post a GIF anyways :)

Rishabh570 commented:
Here is GIF showing a separate list of contributors on both the pages...


![ezgif com-optimize 1](https://user-images.githubusercontent.com/25483260/46648434-b8cf2300-cbb2-11e8-885b-e28cff89c625.gif)

jywarren commented:
Fantastic! I think in testing it i blew through my API limit :-P

I'll move to #53 for that though. (including error for ref)

GREAT WORK again!

```
Failed to load resource: the server responded with a status of 403 (Forbidden)
:8000/#r=all:1 Uncaught (in promise) StatusCodeErrorerror: {message: "API rate limit exceeded for 113.110.231.235. (But …t. Check out the documentation for more details.)", documentation_url: "https://developer.github.com/v3/#rate-limiting"}message: "403 - [object Object]"name: "StatusCodeError"options: {baseUrl: "https://api.github.com", headers: {…}, json: true, method: "GET", qs: {…}, …}response: exports.IncomingMessage {_readableState: ReadableState, readable: false, _events: {…}, _eventsCount: 4, _maxListeners: undefined, …}statusCode: 403isOperational: truestack: "StatusCodeError: 403 - [object Object]↵    at new StatusCodeError (http://localhost:8000/dist/community-toolbox.js:63719:15)↵    at Request.RP$callback [as _callback] (http://localhost:8000/dist/community-toolbox.js:63800:32)↵    at Request.self.callback (http://localhost:8000/dist/community-toolbox.js:65543:22)↵    at emitTwo (http://localhost:8000/dist/community-toolbox.js:31148:13)↵    at Request.emit (http://localhost:8000/dist/community-toolbox.js:31219:7)↵    at Request.<anonymous> (http://localhost:8000/dist/community-toolbox.js:66519:10)↵    at emitOne (http://localhost:8000/dist/community-toolbox.js:31138:13)↵    at Request.emit (http://localhost:8000/dist/community-toolbox.js:31216:7)↵    at exports.IncomingMessage.<anonymous> (http://localhost:8000/dist/community-toolbox.js:66441:12)↵    at Object.onceWrapper (http://localhost:8000/dist/community-toolbox.js:31319:30)↵    at emitNone (http://localhost:8000/dist/community-toolbox.js:31133:20)↵    at exports.IncomingMessage.emit (http://localhost:8000/dist/community-toolbox.js:31213:7)↵    at endReadableNT (http://localhost:8000/dist/community-toolbox.js:62592:12)↵    at afterTickTwo (http://localhost:8000/dist/community-toolbox.js:59159:10)↵    at Item.run (http://localhost:8000/dist/community-toolbox.js:59332:14)↵    at drainQueue (http://localhost:8000/dist/community-toolbox.js:59302:42)"__proto__: Error
```

jywarren commented:
I think the new code uses the limit really fast so the "exhausted" message and localstorage will be much needed. 

BTW we'd love to thank you on Twitter if you wanted to share your handle on there? 🎉

Rishabh570 commented:
> "exhausted" message and localstorage will be much needed.

yes, I do believe that...that's why storing the data is the only best option :)

> BTW we'd love to thank you on Twitter if you wanted to share your handle on there?

Thats so nice :smile_cat: It is *rishabh_570*
